modules/apps/document-library/document-library-document-conversion/src/main/java/com/liferay/document/library/document/conversion/internal/DocumentConversionImpl.java

```diff
@@ -133,19 +140,82 @@ else if (!inputDocumentFormat.isExportableTo(outputDocumentFormat)) {
 		UnsyncByteArrayOutputStream unsyncByteArrayOutputStream =
 			new UnsyncByteArrayOutputStream();
 
-		DocumentConverter documentConverter = _getDocumentConverter();
+		AtomicBoolean isDone = new AtomicBoolean(false);
+		long start = System.currentTimeMillis();
 
-		documentConverter.convert(
-			inputStream, inputDocumentFormat, unsyncByteArrayOutputStream,
-			outputDocumentFormat);
+		long timeout = Math.max(
+			PropsValues.DL_FILE_ENTRY_PREVIEW_GENERATION_TIMEOUT_GHOSTSCRIPT,
+			PropsValues.DL_FILE_ENTRY_PREVIEW_GENERATION_TIMEOUT_PDFBOX);
 
-		FileUtil.write(
-			file, unsyncByteArrayOutputStream.unsafeGetByteArray(), 0,
-			unsyncByteArrayOutputStream.size());
+		_noticeableExecutorService = _portalExecutorManager.getPortalExecutor(
+			DocumentConversionImpl.class.getName());
 
-		inputStream.close();
+		InputStream finalInputStream = inputStream;
 
-		return file;
+		Future<?> future = _noticeableExecutorService.submit(
+			() -> {
+				try {
+					DocumentConverter documentConverter =
+						_getDocumentConverter();
+
+					documentConverter.convert(
+						finalInputStream, inputDocumentFormat,
+						unsyncByteArrayOutputStream, outputDocumentFormat);
+
+					FileUtil.write(
+						file, unsyncByteArrayOutputStream.unsafeGetByteArray(),
+						0, unsyncByteArrayOutputStream.size());
+
+					isDone.set(true);
+
+					if (_log.isInfoEnabled()) {
+						_log.info(
+							StringBundler.concat(
+								"Conversion from ",
+								inputDocumentFormat.getName(), " to ",
+								outputDocumentFormat.getName(), " in ",
+								System.currentTimeMillis() - start, " ms"));
+					}
+				}
+				catch (IOException ioException) {
+					throw new RuntimeException(ioException);
+				}
+				finally {
+					try {
+						finalInputStream.close();
+					}
+					catch (IOException ioException) {
+						_log.error("Unable to close input stream", ioException);
+					}
+				}
+			});
+
+		try {
+			future.get(timeout, TimeUnit.SECONDS);
+
+			futures.put(id, future);
```

Comment:
because on one of my develop version I was doing things with the futures and I did not delete all of it on the final version.

Thanks for pointing at it!